kratos/utilities/geometry_utilities/brep_sbm_utilities.cpp

```diff
@@ -0,0 +1,145 @@
+    if (rSurrogateOuterLoopGeometries.size() > 0) {is_outer_loop_defined = true;}
+    
+    // Loop over rSurrogateOuterLoopGeometries and collect the vertical conditions on each y-knot span
+    for (auto i_cond : rSurrogateOuterLoopGeometries) {
+        auto p_geometry = i_cond;
+        if (std::abs((*p_geometry)[0].X()-(*p_geometry)[1].X()) < 1e-13 ) {
+            // When local refining is performed, the surrogate steps might be larger that the refined knot spans
+            const double direction = ((*p_geometry)[1].Y()-(*p_geometry)[0].Y())/std::abs(((*p_geometry)[1].Y()-(*p_geometry)[0].Y()));
+
+            int i_row_1 = FindKnotSpans1D(rSpansV, (*p_geometry)[0].Y() + 1e-12*(direction));
+            int i_row_2 = FindKnotSpans1D(rSpansV, (*p_geometry)[1].Y() - 1e-12*(direction));
+
+            for (int i_row = std::min(i_row_1, i_row_2); i_row < std::max(i_row_1,i_row_2)+1; i_row++) {
+                vertical_conditions_per_row[i_row].push_back((*p_geometry)[0].X());
+            }
+        }
+    }
+    
+    // Loop over rSurrogateInnerLoopGeometries and collect the vertical conditions on each y-knot span
+    for (auto i_cond : rSurrogateInnerLoopGeometries) {
+        auto p_geometry = i_cond; 
+        if (std::abs((*p_geometry)[0].X()-(*p_geometry)[1].X()) < 1e-13 ) {
+            // When local refining is performed, the surrogate steps might be larger that the refined knot spans
+            const double direction = ((*p_geometry)[1].Y()-(*p_geometry)[0].Y())/std::abs(((*p_geometry)[1].Y()-(*p_geometry)[0].Y()));
+            int i_row_1 = FindKnotSpans1D(rSpansV, (*p_geometry)[0].Y() + 1e-12*(direction));
+            int i_row_2 = FindKnotSpans1D(rSpansV, (*p_geometry)[1].Y() - 1e-12*(direction));
+
+            for (int i_row = std::min(i_row_1, i_row_2); i_row < std::max(i_row_1,i_row_2)+1; i_row++) {
+                vertical_conditions_per_row[i_row].push_back((*p_geometry)[0].X());
+            }
+        }
+    }
+
+    // Sort the vertical conditions by the x coordinate
+    for (IndexType i_row = 0; i_row < rSpansV.size() - 1; ++i_row) {
+        std::sort(vertical_conditions_per_row[i_row].begin(), vertical_conditions_per_row[i_row].end());  
+    }   
+    
+    // Loop over each row and creation of the integration points
+    for (IndexType j = 0; j < rSpansV.size() - 1; j++) {
+        IndexType starting_column_index = 0;
+        IndexType next_switch_knot_span;
+        if (vertical_conditions_per_row[j].size() > 0) {
+            next_switch_knot_span = FindKnotSpans1D(rSpansU, vertical_conditions_per_row[j][0]+1e-13);
+            }
+        else {
+            next_switch_knot_span = rSpansU.size() - 1;
+        }
+        IndexType vertical_condition_index = 1;
+
+        if (is_outer_loop_defined){
+            if (vertical_conditions_per_row[j].empty()) {continue;}
+            starting_column_index = FindKnotSpans1D(rSpansU, vertical_conditions_per_row[j][0]+1e-13);
```

Comment:
```suggestion
            starting_column_index = FindKnotSpans1D(rSpansU, vertical_conditions_per_row[j][0]+std::numeric_limits<double>::epsilon());
```
I guess you need this. Same below.